blog/_posts/2022-04-04-a-dev-environment-repl.md

```diff
@@ -0,0 +1,132 @@
+
+Tilt, as a tool for assembling development environments, is all about tight [feedback loops][controlloop]. Tilt responds to changes in your environment in the same way that the REPL snaps to action when you hit ENTER. One of the events Tilt responds to is changes to the `Tiltfile`, Tilt's own [Starlark][starlark]-based configuration file. With a little creativity and flexibility, we can run Tilt in a terminal inside or next to our editor (`tilt up --stream`) and mimic the feedback and responsiveness of a REPL:
+
+![Tilt Fibonacci example](/assets/images/a-dev-environment-repl/tilt-fib.gif)
+
+The result is a loop that feels surprisingly responsive: write a bit of code, hit `Ctrl-S` or `⌘-S`, and see the result! If you're exploring Tilt and haven't used Starlark (or Python, Starlark's parent language), this can be a great way to fool around and get your bearings with tiltfile syntax or builtins.
+
+You may have noticed Tilt complaining at the end of the loop about "No resources found". We implemented and ran a Fibonacci function in Starlark, but didn't give Tilt anything else to do. Tilt primarily wants to build and run things for you; Starlark is the flexible medium through which you describe what to run and build. Tilt is that eager puppy that will keep fetching the ball, though finding it unsatisfying and wishing it was a bone to chew on instead. 
+
+It's at this point in the experimentation phase where it helps to view Tilt as a [replacement for Bash](/2018/12/05/tilt-is-the-start-sh-script-of-my-dreams.html) or your command-line shell of choice. Think about activities you normally would type into your shell and see if you can wire them into your tiltfile and have Tilt run them for you. For example, maybe you're sitting down at the keyboard at the beginning of your work day and would normally use `git` to pull new changes to your code. Instead, create a `local_resource` to do it for you:
+
+```python
+local_resource("update-code", "git pull")
+```
+
+Tilt responds with:
+
+```
+Initial Build
+Loading Tiltfile at: /Users/nicksieger/tilt-dev/tilt-avatars/web/Tiltfile
+Successfully loaded Tiltfile (413.333µs)
+  update-code │
+  update-code │ Initial Build
+  update-code │ Running cmd: git pull
+  update-code │ Already up to date.
+```
+
+Now Tilt will pull code for you when you `tilt up`. [Resources][resources] are Tilt's unit of work; in the case of a `local_resource`, it's a command to run on your local machine. Resources are stateful in that Tilt looks for changes to them every time the tiltfile is executed. So you can also experiment and iterate on resources; if Tilt detects relevant differences, it will re-execute the resource. Say that you find that `git pull` is not quite what you want Tilt to do; instead, you want some conditional logic to check if you have a clean working copy and only pull new changes then. You can change the `update-code` resource, all while Tilt is running:
```

Comment:
```suggestion
Now Tilt will pull code for you when you `tilt up`. [Resources][resources] are Tilt's unit of work; in the case of a `local_resource`, it's a command to run on your local machine. Resources are stateful in that Tilt looks for changes to them every time the `Tiltfile` is executed. So you can also experiment and iterate on resources; if Tilt detects relevant differences, it will re-execute the resource. Say that you find that `git pull` is not quite what you want Tilt to do; instead, you want some conditional logic to check if you have a clean working copy and only pull new changes then. You can change the `update-code` resource, all while Tilt is running:
```